torch/utils/checkpoint.py

```diff
@@ -159,31 +162,31 @@ def add_device_types(arg):
 # the device of all Tensor args.
 #
 # To consider:  maybe get_device_states and set_device_states should reside in torch/random.py?
-def get_device_states(*args) -> Tuple[List[int], List[torch.Tensor]]:
+def get_device_states(*args) -> Tuple[List[torch.device], List[torch.Tensor]]:
     # This will not error out if "arg" is a CPU tensor or a non-tensor type because
     # the conditionals short-circuit.
-    fwd_device_ids = []
+    fwd_devices = []
 
-    def add_device_ids(arg):
-        nonlocal fwd_device_ids
+    def add_device(arg):
+        nonlocal fwd_devices
         if isinstance(arg, torch.Tensor) and not arg.device.type == "cpu":
-            fwd_device_ids.append(arg.get_device())
-    tree_map(add_device_ids, args)
+            fwd_devices.append(arg.device)
+    tree_map(add_device, args)
 
     fwd_device_states = []
     device_module = _get_device_module(_infer_device_type(*args))
 
-    for device_id in fwd_device_ids:
-        with device_module.device(device_id):
+    for device in fwd_devices:
+        with device_module.device(device):
             fwd_device_states.append(device_module.get_rng_state())
 
-    return fwd_device_ids, fwd_device_states
+    return fwd_devices, fwd_device_states
 
 
 def set_device_states(devices, states) -> None:
-    device_module = _get_device_module(_infer_device_type(*states))
+    device_module = _get_device_module(_infer_device_type_from_devices(devices))
```

Comment:
`_get_device_module` supports different devices and `_infer_device_type` also supports different devices. May I know why we need to such changes?

Reply:
`_get_device_module` returns appropriate module (`torch.cuda` or `torch.xpu`) depending on the given device type.

`_infer_device_type` returns single device type on which given tensors are allocated (these are rng_state tensors). If tensors are on different devices, one will be returned with the warning printed. See logic in the function.

The whole problem and a bug is that rng_state tensors are CPU tensors and does not carry device information with them at all as far as I can tell. So, `_infer_device_type` will  always return `cuda` device type since it's the default for the case when all tensors are cpu ones (again - see logic in the function).

My change is to deduce device type not from the rng_state tensors, but from devices list saved on `forward()`. Previously this information was just discarded and only device index was saved.